fem/fespace.hpp

```diff
@@ -609,44 +597,206 @@ class QuadratureSpace
+    using interpolation. */
+/** The forward, coarse-to-fine, transfer uses nodal interpolation. The
+    backward, fine-to-coarse, transfer is defined locally (on a coarse element)
+    as B = (F^t M_f F)^{-1} F^t M_f, where F is the forward transfer matrix, and
+    M_f is a mass matrix on the union of all fine elements comprising the coarse
+    element. Note that the backward transfer operator, B, is a left inverse of
+    the forward transfer operator, F, i.e. B F = I. Both F and B are defined in
+    reference space and do not depend on the actual physical shape of the mesh
+    elements.
+
+    It is assumed that both the coarse and the fine FiniteElementSpace%s use
+    compatible types of elements, e.g. finite elements with the same map-type
+    (VALUE, INTEGRAL, H_DIV, H_CURL - see class FiniteElement). Generally, the
+    FE spaces can have different orders, however, in order for the backward
+    operator to be well-defined, the (local) number of the fine dofs should not
+    be smaller than the number of coarse dofs. */
+class InterpolationGridTransfer : public GridTransfer
 {
-   const L2Projection &l2proj;
+protected:
+   BilinearFormIntegrator *mass_integ; ///< Ownership depends on #own_mass_integ
+   bool own_mass_integ; ///< Ownership flag for #mass_integ
+
+   OperatorHandle F; ///< Forward, coarse-to-fine, operator
+   OperatorHandle B; ///< Backward, fine-to-coarse, operator
+
 public:
-   L2Prolongation(const L2Projection &l2proj_) : l2proj(l2proj_) { }
-   void Mult(const Vector &x, Vector &y) const
+   InterpolationGridTransfer(FiniteElementSpace &coarse_fes,
+                             FiniteElementSpace &fine_fes)
+      : GridTransfer(coarse_fes, fine_fes),
+        mass_integ(NULL), own_mass_integ(false)
+   { }
+
+   virtual ~InterpolationGridTransfer();
+
+   /** @brief Assign a mass integrator to be used in the construction of the
+       backward, fine-to-coarse, transfer operator. */
+   void SetMassIntegrator(BilinearFormIntegrator *mass_integ_,
+                          bool own_mass_integ_ = true);
+
+   virtual const Operator &ForwardOperator();
+
+   virtual const Operator &BackwardOperator();
+};
+
+
+/** @brief Transfer data between a coarse mesh and an embedded refined mesh
+    using L2 projection. */
+/** The forward, coarse-to-fine, transfer uses L2 projection. The backward,
+    fine-to-coarse, transfer is defined locally (on a coarse element) as
+    B = (F^t M_f F)^{-1} F^t M_f, where F is the forward transfer matrix, and
+    M_f is the mass matrix on the union of all fine elements comprising the
+    coarse element. Note that the backward transfer operator, B, is a left
+    inverse of the forward transfer operator, F, i.e. B F = I. Both F and B are
+    defined in physical space and, generally, vary between different mesh
+    elements.
+
+    This class currently supports only L2 finite element spaces and fine meshes
```

Comment:
The L2 transfer operator will "work" even on H1 finite element spaces (it just won't have the desired conservation properties). It won't completely break though.

Reply:
Oh, okay, I did not realize that. Please update the documentation to reflect that.

Reply:
Ok, I added some explanatory comments.